nbgrader/auth/hubauth.py

```diff
@@ -0,0 +1,123 @@
+        response = self._proxy_request('/api/routes/hub/formgrade', method='POST', body={
+            'target': self._base_url
+        })
+        if response.status_code != 201:
+            raise Exception('Error while trying to add JupyterHub route. {}: {}'.format(response.status_code, response.text))
+
+        # Redirect all formgrade request to the correct API method.
+        self._app.register_blueprint(blueprint, static_url_path='/hub/formgrade/static', url_prefix='/hub/formgrade', url_defaults={'name': 'hub'})
+
+    def authenticate(self):
+        """Authenticate a request.
+        Returns a boolean or flask redirect."""
+
+        # If auth cookie doesn't exist, redirect to the login page with
+        # next set to redirect back to the this page.
+        if 'jupyter-hub-token' not in request.cookies:
+            return redirect(self._hub_base_url + '/hub/login?next=' + self._hub_base_url + '/hub/formgrade')
+        cookie = request.cookies[self.hubapi_cookie]
+
+        # Check with the Hub to see if the auth cookie is valid.
+        response = self._hubapi_request('/hub/api/authorizations/cookie/' + self.hubapi_cookie, body=cookie)
+        if response.status_code == 200:
+
+            #  Auth information recieved.
+            data = response.json()
+            if 'user' in data:
+                user = data['user']
+
+                # Check if the user name is registered as a grader.
+                print(self.graders)
+                print(user)
+                if user in self.graders:
+                    return True
+            else:
+                abort(500, "Failed to check authorization, malformed response from Hub auth.")
+        elif response.status_code == 403:
+            self.log.error("I don't have permission to verify cookies, my auth token may have expired: [%i] %s", response.status_code, response.reason)
+            abort(500, "Permission failure checking authorization, I may need to be restarted")
+        elif response.status_code >= 500:
+            self.log.error("Upstream failure verifying auth token: [%i] %s", response.status_code, response.reason)
+            abort(502, "Failed to check authorization (upstream problem)")
+        elif response.status_code >= 400:
+            self.log.warn("Failed to check authorization: [%i] %s", response.status_code, response.reason)
+            abort(500, "Failed to check authorization")
+        else:
+            # Auth invalid, reauthenticate.
+            return redirect(self._hub_base_url + '/hub/login?next=' + self._hub_base_url + '/hub/formgrade')
+        return False
+
+    def notebook_server_exists(self):
+        """Does the notebook server exist?"""
+        return True
+
+    def get_notebook_url(self, relative_path):
+        """Gets the notebook's url."""
+        # TODO
+        return "http://{}:{}/notebooks/{}".format(
+            self.hub_address,
+            self.hub_port,
+            relative_path)
```

Comment:
Ah yes, you're right.  It should be /user/notebooks.  I haven't tested this with notebooks yet, but will test today.